File: src/main/java/com/heekng/celloct/entity/Work.java
package com.heekng.celloct.entity;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.time.LocalDate;
import java.time.LocalDateTime;

import static javax.persistence.CascadeType.*;
import static javax.persistence.FetchType.*;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Table(
        uniqueConstraints = {
                @UniqueConstraint(
                        name = "work_work_date_staff_id_unique",
                        columnNames = {"work_date", "staff_id"}
                )
        }
)
public class Work extends BaseTimeEntity {

    @Id @GeneratedValue
    @Column(name = "work_id")
    private Long id;

    @Embedded
    private WorkTime workTime;

    @Column(name = "note")
    private String note;

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "staff_id", nullable = false)
    private Staff staff;

    @OneToOne(mappedBy = "work", fetch = LAZY, cascade = REMOVE)
    @JoinColumn(name = "work_update_request_id")
    private WorkUpdateRequest workUpdateRequest;

    @Builder
    public Work(LocalDate workDate, Staff staff, LocalDateTime startDate, LocalDateTime endDate, String note) {
        this.staff = staff;
        this.note = note;
        this.workTime = WorkTime.builder()
                .workDate(workDate)
                .startDate(startDate)
                .endDate(endDate)
                .build();
    }

    public void addWorkUpdateRequest(WorkUpdateRequest workUpdateRequest) {
        this.workUpdateRequest = workUpdateRequest;
    }

    public void deleteWorkUpdateRequest() {
        this.workUpdateRequest = null;
    }
}

File: src/main/kotlin/com/heekng/celloct/entity/Work.kt
package com.heekng.celloct.entity

import java.time.LocalDateTime
import javax.persistence.*

@Entity
@Table(
    uniqueConstraints = [UniqueConstraint(
        name = "work_work_date_staff_id_unique",
        columnNames = ["work_date", "staff_id"]
    )]
)
class Work(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY) @Column(name = "work_id")
    val id: Long? = null,

    @Embedded
    val workTime: WorkTime,

    @Column(name = "note")
    var note: String? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "staff_id", nullable = false)
    val staff: Staff,

    @OneToOne(mappedBy = "work", fetch = FetchType.LAZY, cascade = [CascadeType.REMOVE])
    @JoinColumn(name = "work_update_request_id")
    var workUpdateRequest: WorkUpdateRequest? = null,
) : BaseTimeEntity() {

    fun addWorkUpdateRequest(workUpdateRequest: WorkUpdateRequest) {
        this.workUpdateRequest = workUpdateRequest
    }

    fun deleteWorkUpdateRequest() {
        this.workUpdateRequest = null
    }

    fun updateWork(
        startDate: LocalDateTime,
        endDate: LocalDateTime,
        note: String? = null,
    ) {
        this.workTime.changeWorkTime(startDate, endDate)
        this.note = note
    }
}
